### PR TITLE
[Internal] Add converter functions and tests for plugin framework

### PR DESCRIPTION
**Pull request by edwardfeng-db**
## Changes

- Addressed comments from https://github.com/databricks/terraform-provider-databricks/pull/3893
  - Created new package for plugin framework and added converter functions between tfsdk and gosdk structs
  - Added unit tests
  - Moved `ListAllFields` from `common` into `internal/reflect_utils`

## Tests

- [x] `make test` run locally
- [x] relevant change in `docs/` folder
- [x] covered with integration tests in `internal/acceptance`
- [x] relevant acceptance tests are passing
- [x] using Go SDK

### CODE REVIEW

File: pluginframework/common/go_to_tf.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+		stringResult := stringMethod.Call(nil)
+		if len(stringResult) == 1 {
+			return stringResult[0].Interface().(string)
+		} else {
+			panic("num get string has more than one result")
+		}
+	} else {
+		panic("enum does not have valid .String() method")
+	}
+}
```
Comment (edwardfeng-db) [inline pluginframework/common/go_to_tf.go:212]: See comment [here](https://github.com/databricks/terraform-provider-databricks/pull/3893/files#r1721764318) for why type casting doesn't necessarily make the code simpler

File: internal/reflect_utils/reflect_utils.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package reflect_utils
+
+import "reflect"
+
+type Field struct {
+	Sf reflect.StructField
```
Comment (mgyucht) [inline internal/reflect_utils/reflect_utils.go:6]: Can we just name this `StructField`? I always forget what `Sf` means

File: internal/reflect_utils/reflect_utils.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	V  reflect.Value
+}
+
+func ListAllFields(v reflect.Value) []Field {
```
Comment (mgyucht) [inline internal/reflect_utils/reflect_utils.go:10]: Can you add a doccomment here?

File: pluginframework/common/converters_test.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
+package pluginframework
```
Comment (mgyucht) [inline pluginframework/common/converters_test.go:1]: Can we move these files into `internal/pluginframework/converters`? Keep functions internal that don't need to be exposed publicly, plus converters is a better package name than common or pluginframework.

File: internal/reflect_utils/reflect_utils.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package reflect_utils
```
Comment (mgyucht) [inline internal/reflect_utils/reflect_utils.go:1]: I hate the name `utils` in packages... see https://go.dev/wiki/CodeReviewComments#package-names. If we don't want to clash with `reflect` from Go's standard library, let's call this `tfreflect` instead.

File: pluginframework/common/converters_test.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+	"reflect"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/stretchr/testify/assert"
+)
+
+type DummyTfSdk struct {
+	Enabled           types.Bool                  `tfsdk:"enabled" tf:"optional"`
+	Workers           types.Int64                 `tfsdk:"workers" tf:""` // Test required field
```
Comment (mgyucht) [inline pluginframework/common/converters_test.go:15]: Is the `tf` option present on every field?

File: pluginframework/common/converters_test.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+type DummyNestedGoSdk struct {
+	Name            string   `json:"name"`
+	Enabled         bool     `json:"enabled"`
+	ForceSendFields []string `json:"-"`
+}
+
+// Function to construct individual test case with a pair of matching tfSdkStruct and gosdkStruct.
+// Verifies that the conversion both ways are working as expected.
+func ConverterTestCase(t *testing.T, description string, tfSdkStruct DummyTfSdk, goSdkStruct DummyGoSdk) {
```
Comment (mgyucht) [inline pluginframework/common/converters_test.go:93]: This is great, thank you for splitting this into different test cases. However, these will all be reported as the same test case.

Instead, let's follow the pattern of table driven tests: https://go.dev/wiki/TableDrivenTests. The general structure will look like this:
```go
var tests = []struct{
  name string
  tfSdkStruct DummyTfSdk
  goSdkStruct DummyGoSdk
}{
  {"case 1", DummyTfSdk{}, DummyGoSdk{}},
  ... more cases
}

func TestConverter(t *testing.T) {
  for _, tc := range tests {
    t.Run(tc.name, func() { ConverterTestCase(t, tc.name, tc.tfSdkStruct, tc.goSdkStruct) })
  }
}
```

File: pluginframework/common/go_to_tf.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+		boolVal := srcFieldValue.(bool)
+		// check if the value is non-zero or if the field is in the forceSendFields list
+		if boolVal || forceSendField {
+			destField.Set(reflect.ValueOf(types.BoolValue(boolVal)))
+		} else {
+			destField.Set(reflect.ValueOf(types.BoolNull()))
+		}
+	case reflect.Int64:
+		// convert any kind of integer to int64
+		intVal := srcField.Convert(reflect.TypeOf(int64(0))).Interface().(int64)
```
Comment (mgyucht) [inline pluginframework/common/go_to_tf.go:115]: Does this work?
```suggestion
		intVal := srcField.Convert(reflect.TypeOf(int64(0))).Int()
```

File: pluginframework/common/go_to_tf.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+		// check if the value is non-zero or if the field is in the forceSendFields list
+		if intVal != 0 || forceSendField {
+			destField.Set(reflect.ValueOf(types.Int64Value(int64(intVal))))
+		} else {
+			destField.Set(reflect.ValueOf(types.Int64Null()))
+		}
+	case reflect.Float64:
+		// convert any kind of float to float64
+		float64Val := srcField.Convert(reflect.TypeOf(float64(0))).Interface().(float64)
```
Comment (mgyucht) [inline pluginframework/common/go_to_tf.go:124]: Does this work?
```suggestion
		float64Val := srcField.Convert(reflect.TypeOf(float64(0))).Float()
```

File: pluginframework/common/go_to_tf.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+			destField.Set(reflect.ValueOf(types.BoolValue(boolVal)))
+		} else {
+			destField.Set(reflect.ValueOf(types.BoolNull()))
+		}
+	case reflect.Int64:
+		// convert any kind of integer to int64
+		intVal := srcField.Convert(reflect.TypeOf(int64(0))).Interface().(int64)
+		// check if the value is non-zero or if the field is in the forceSendFields list
+		if intVal != 0 || forceSendField {
+			destField.Set(reflect.ValueOf(types.Int64Value(int64(intVal))))
```
Comment (mgyucht) [inline pluginframework/common/go_to_tf.go:118]: Unnecessary cast?
```suggestion
			destField.Set(reflect.ValueOf(types.Int64Value(intVal)))
```

File: pluginframework/common/go_to_tf.go [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+		} else {
+			destField.Set(reflect.ValueOf(types.StringNull()))
+		}
+	case reflect.Struct:
+		if srcField.IsZero() {
+			// Skip zeros
+			return nil
+		}
+		// resolve the nested struct by recursively calling the function
+		if GoSdkToTfSdkStruct(srcFieldValue, destField.Addr().Interface(), ctx).ErrorsCount() > 0 {
```
Comment (mgyucht) [inline pluginframework/common/go_to_tf.go:151]:
```suggestion
		if GoSdkToTfSdkStruct(srcFieldValue, destField.Addr().Interface(), ctx).HasErrors() {
```

File: pluginframework/common/go_to_tf.go [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+			strVal = srcFieldValue.(string)
+		}
+		// check if the value is non-zero or if the field is in the forceSendFields list
+		if strVal != "" || forceSendField {
+			destField.Set(reflect.ValueOf(types.StringValue(strVal)))
+		} else {
+			destField.Set(reflect.ValueOf(types.StringNull()))
+		}
+	case reflect.Struct:
+		if srcField.IsZero() {
```
Comment (mgyucht) [inline pluginframework/common/go_to_tf.go:146]: Do we need to worry about ForceSendFields here?

File: pluginframework/common/go_to_tf.go [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+			// Skip zeros
+			return nil
+		}
+		// resolve the nested struct by recursively calling the function
+		if GoSdkToTfSdkStruct(srcFieldValue, destField.Addr().Interface(), ctx).ErrorsCount() > 0 {
+			panic("Error converting gosdk to tfsdk struct")
+		}
+	case reflect.Slice:
+		if srcField.IsNil() {
```
Comment (mgyucht) [inline pluginframework/common/go_to_tf.go:155]: Here we don't need to worry about FSF because our API can't distinguish between no list and empty list in requests.

File: pluginframework/common/go_to_tf.go (hunk 1 above)
Comment (mgyucht) [reply pluginframework/common/go_to_tf.go:213]: Did we ever figure out why this is?

File: pluginframework/common/go_to_tf.go [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+			return diag.Diagnostics{diag.NewErrorDiagnostic(err.Error(), "gosdk to tfsdk field conversion failure")}
+		}
+	}
+	return nil
+}
+
+func goSdkToTfSdkSingleField(srcField reflect.Value, destField reflect.Value, forceSendField bool, ctx context.Context) error {
+
+	if !destField.CanSet() {
+		panic(fmt.Errorf("destination field can not be set: %s", destField.Type().Name()))
```
Comment (mgyucht) [inline pluginframework/common/go_to_tf.go:84]: Anywhere that we panic, we should print an error that says something like "This is a bug. Please report this to the maintainers at github.com/databricks/terraform-provider-databricks."

File: pluginframework/common/tf_to_go.go [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+
+		destField := destVal.FieldByName(srcFieldName)
+
+		err := tfSdkToGoSdkSingleField(srcField, destField, srcFieldName, &forceSendFieldsField, ctx)
+		if err != nil {
+			return diag.Diagnostics{diag.NewErrorDiagnostic(err.Error(), "tfsdk to gosdk field conversion failure")}
+		}
+	}
+
+	return diag.Diagnostics{}
```
Comment (mgyucht) [inline pluginframework/common/tf_to_go.go:67]: return nil?

File: pluginframework/common/tf_to_go.go [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+			if srcField.IsZero() {
+				return
+			}
+
+			// Find the Set method
+			destVal := reflect.New(destField.Type())
+			setMethod := destVal.MethodByName("Set")
+			if !setMethod.IsValid() {
+				panic(fmt.Sprintf("set method not found on enum type: %s", destField.Type().Name()))
+			}
```
Comment (mgyucht) [inline pluginframework/common/tf_to_go.go:170]: We don't have the same case with enum pointers as above?

**Review by mgyucht — COMMENTED** [review]
Looking great, few comments but this is much nicer!

File: pluginframework/common/converters_test.go (hunk 6 above)
Comment (edwardfeng-db) [reply pluginframework/common/converters_test.go:15]: Yes

File: pluginframework/common/tf_to_go.go [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+//	types.Float64 -> float64
+//	types.String -> string
+//
+// NOTE:
+//
+//	ForceSendFields are populated for string, bool, int64, float64 on non null values
+//	types.list and types.map are not supported
+//	map keys should always be a string
+//	tfsdk structs use types.String for all enum values
+func TfSdkToGoSdkStruct(tfsdk interface{}, gosdk interface{}, ctx context.Context) diag.Diagnostics {
```
Comment (mgyucht) [inline pluginframework/common/tf_to_go.go:30]: ctx should be first argument

File: pluginframework/common/tf_to_go.go [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+		// If it is a real stuct instead of a tfsdk type, recursively resolve it.
+		if TfSdkToGoSdkStruct(srcFieldValue, destField.Addr().Interface(), ctx).ErrorsCount() > 0 {
+			panic("Error converting tfsdk to gosdk struct")
+		}
+	}
+}
+
+func addToForceSendFields(fieldName string, forceSendFieldsField *reflect.Value) {
+	if forceSendFieldsField == nil || !forceSendFieldsField.IsValid() || !forceSendFieldsField.CanSet() {
+		log.Printf("[Debug] forceSendFieldsField is nil, invalid or not settable. %s", fieldName)
```
Comment (mgyucht) [inline pluginframework/common/tf_to_go.go:208]: Use `tflog.Debug`

File: pluginframework/common/tf_to_go.go (hunk 16 above)
Comment (mgyucht) [reply pluginframework/common/tf_to_go.go:170]: Not needed because we always instantiate the memory for this field in the same way two lines above.

File: pluginframework/common/go_to_tf.go (hunk 12 above)
Comment (edwardfeng-db) [reply pluginframework/common/go_to_tf.go:146]: Not really - discussed offline, this is actuallly not supported by the gosdk, so we don't need FSF

File: pluginframework/common/go_to_tf.go (hunk 1 above)
Comment (edwardfeng-db) [reply pluginframework/common/go_to_tf.go:213]: Discussed offline and added comment, this case where `CanAddr()` returns false happens in unit tests due to the value being a constant

File: internal/tfreflect/reflect_utils.go [hunk 19]
```diff
@@ -0,0 +1,27 @@
+package tfreflect
+
+import "reflect"
+
+type Field struct {
+	StructField reflect.StructField
+	Value       reflect.Value
+}
+
+// Given a reflect.Value of a struct, list all of the fields for both struct field and
```
Comment (mgyucht) [inline internal/tfreflect/reflect_utils.go:10]: Small nit:
> Comments should begin with the name of the thing being described and end in a period.

https://go.dev/wiki/CodeReviewComments#comment-sentences

File: pluginframework/converters/converters_test.go [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+	name        string
+	tfSdkStruct DummyTfSdk
+	goSdkStruct DummyGoSdk
+}{
+	{
+		"string conversion",
+		DummyTfSdk{Description: types.StringValue("abc")},
+		DummyGoSdk{Description: "abc", ForceSendFields: []string{"Description"}},
+	},
+	{
```
Comment (mgyucht) [inline pluginframework/converters/converters_test.go:113]: Can we add one test case for DummyTfSdk value set to zero values for each of the primitive types and verifying that ForceSendFields is set properly?

File: pluginframework/converters/go_to_tf.go [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+			strVal = getStringFromEnum(srcField)
+		} else {
+			strVal = srcFieldValue.(string)
+		}
+		// check if the value is non-zero or if the field is in the forceSendFields list
+		if strVal != "" || forceSendField {
+			destField.Set(reflect.ValueOf(types.StringValue(strVal)))
+		} else {
+			destField.Set(reflect.ValueOf(types.StringNull()))
+		}
```
Comment (mgyucht) [inline pluginframework/converters/go_to_tf.go:145]: Is the else case needed? The default value should be empty string IIUC.

File: pluginframework/converters/go_to_tf.go [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+
+	"github.com/databricks/databricks-sdk-go/logger"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+
+	"github.com/databricks/terraform-provider-databricks/common"
+	"github.com/databricks/terraform-provider-databricks/internal/tfreflect"
+)
+
+// Converts a gosdk struct into a tfsdk struct, with the folowing rules.
```
Comment (mgyucht) [inline pluginframework/converters/go_to_tf.go:16]: Ditto here, just put the function name first.

File: pluginframework/converters/tf_to_go.go [hunk 23]
```diff
@@ -0,0 +1,221 @@
...
+	"github.com/databricks/databricks-sdk-go/logger"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
+	"github.com/databricks/terraform-provider-databricks/common"
+	"github.com/databricks/terraform-provider-databricks/internal/tfreflect"
+)
+
+// Converts a tfsdk struct into a gosdk struct, with the folowing rules.
```
Comment (mgyucht) [inline pluginframework/converters/tf_to_go.go:17]: Ditto here, just put the function name first.

**Review by mgyucht — APPROVED** [review]
LGTM, with a few nits.

File: pluginframework/converters/converters_test.go (hunk 20 above)
Comment (edwardfeng-db) [reply pluginframework/converters/converters_test.go:113]: Sounds good, added

File: pluginframework/converters/go_to_tf.go (hunk 21 above)
Comment (edwardfeng-db) [reply pluginframework/converters/go_to_tf.go:145]: Setting `StringNull` just means that we are explicitly saying this is the case that is "unset" thus when we convert it back to gosdk struct, we know whether to add it into FSF or not
